allennlp/nlp_api.py

```diff
@@ -0,0 +1,289 @@
+            name that was a key in the ``embedders`` dictionary passed to the constructor of
+            ``NlpApi``.  If the name is not present in that dictionary, the behavior is determined
+            by the ``fallback_behavior`` parameter.
+        fallback_behavior : ``str``, optional (default=``"crash"``)
+            Determines what to do when ``name`` is not a key in ``embedders``.  There are two
+            options:
+
+            - ``"crash"``: raise an error.  This is the default.  The intention is to help you find
+              bugs - if you specify a particular embedding name when constructing your model
+              without giving a fallback behavior, you probably wanted to use a particular set of
+              modules, so we crash if they are not provided.
+            - ``"use [name]``: In this case, we `reuse` the token embedders with the given name,
+              essentially replacing the ``name`` given as an argument to this function with the
+              name in the fallback behavior.  You should use this fallback behavior to share
+              embedding layers across inputs.
+
+            An alternative usage of this API to the BiDAF example given above, that allows for more
+            flexibility in using different embedded representations for different fields, would be
+            something like this::
+
+                def forward(self, inputs):
+                    passage_tokens = inputs['passage']
+                    embedded_passage = nlp_api.embed_text_field(passage_tokens,
+                                                                "passage",
+                                                                fallback_behavior="use default")
+                    question_tokens = inputs['question']
+                    embedded_question = nlp_api.embed_text_field(question_tokens,
+                                                                 "question",
+                                                                 fallback_behavior="use passage")
+
+            This would let you specify `different` embedding behaviors for the "passage" and
+            "question" fields in your configuration file, but still falling back to using a shared
+            default embedding if no special configuration was given for "passage" or "question"
+            embeddings.
+        """
+        if name not in self._embedders:
+            if fallback_behavior == "crash":
+                raise ConfigurationError("No embedders specified for name: %s" % name)
+            elif fallback_behavior.startswith("use "):
+                name_to_use = fallback_behavior[4:]
+                self._embedders[name] = self._embedders[name_to_use]
+            else:
+                raise ConfigurationError("Unrecognized fallback behavior: %s" % fallback_behavior)
+        embedders = self._embedders[name]
+
+        if isinstance(embedders, torch.nn.Module):
```

Comment:
The reason you need it this way is to handle the case where we want to use a separate model to do the embedding, like in @matt-peters' work.  In that case, you might have several input token arrays, but only one embedder (because it's a pretrained model that will handle all of the arrays itself).  We really want that to have a clearly different signature here, so we can distinguish between something that takes all inputs, no matter how many there are, and just having one input array.